// meteor-immutable-observer/lib/ImmutableMapObserver.js
'use strict';

exports.__esModule = true;
exports['default'] = ImmutableMapObserver;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _immutable = require('immutable');

var _immutable2 = _interopRequireDefault(_immutable);

var _updateDeep = require('./updateDeep');

var _updateDeep2 = _interopRequireDefault(_updateDeep);

function mergeChanges(document, fields) {
  return document.withMutations(function (document) {
    for (var key in fields) {
      if (fields.hasOwnProperty(key)) {
        var newValue = fields[key];
        if (newValue === undefined) {
          document['delete'](key);
        } else {
          document.update(key, function (oldValue) {
            return _updateDeep2['default'](oldValue, _immutable2['default'].fromJS(newValue));
          });
        }
      }
    }
  });
}

function ImmutableMapObserver(cursor) {
  if (Tracker.active) {
    throw new Error("This can't be used inside reactive computations; it could cause infinite invalidate loops");
  }

  var _documents = undefined;
  var dep = new Tracker.Dependency();

  function update(newDocuments) {
    _documents = newDocuments;
    dep.changed();
  }

  var initialDocuments = {};
  var handle = cursor.observeChanges({
    added: function added(id, fields) {
      fields._id = id;
      if (initialDocuments) {
        initialDocuments[id] = _immutable2['default'].fromJS(fields);
      } else {
        update(_documents.set(id, _immutable2['default'].fromJS(fields)));
      }
    },
    changed: function changed(id, fields) {
      update(_documents.update(id, function (document) {
        return mergeChanges(document, fields);
      }));
    },
    removed: function removed(id) {
      update(_documents['delete'](id));
    }
  });
  _documents = _immutable2['default'].Map(initialDocuments);
  initialDocuments = undefined;

  return {
    documents: function documents() {
      dep.depend();
      return _documents;
    },
    stop: function stop() {
      handle.stop();
    }
  };
}

module.exports = exports['default'];
